spring-cloud-stream/src/main/java/org/springframework/cloud/stream/converter/JavaSerializationMessageConverter.java

```diff
@@ -16,44 +16,51 @@
 
 package org.springframework.cloud.stream.converter;
 
+import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
-import java.io.Serializable;
+import java.util.Arrays;
 
 import org.springframework.messaging.Message;
-
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.AbstractMessageConverter;
 
 /**
- * A {@link org.springframework.messaging.converter.MessageConverter}
- * to convert from a POJO to byte[] with Java.io serialization if any.
- *
- * @author David Turanski
+ * @author Marius Bogoevici
  */
-public class JavaToSerializedMessageConverter extends AbstractFromMessageConverter {
+public class JavaSerializationMessageConverter extends AbstractMessageConverter {
 
-	public JavaToSerializedMessageConverter() {
-		super(MessageConverterUtils.X_JAVA_OBJECT, MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT);
+	public JavaSerializationMessageConverter() {
+		super(Arrays.asList(MessageConverterUtils.X_JAVA_SERIALIZED_OBJECT, MessageConverterUtils.X_JAVA_OBJECT));
 	}
 
 	@Override
-	protected Class<?>[] supportedTargetTypes() {
-		return new Class<?>[] { byte[].class };
+	protected boolean supports(Class<?> clazz) {
+		return true;
 	}
 
 	@Override
-	protected Class<?>[] supportedPayloadTypes() {
-		return new Class<?>[] { Serializable.class };
+	public Object convertFromInternal(Message<?> message, Class<?> targetClass, Object conversionHint) {
+		ByteArrayInputStream bis = new ByteArrayInputStream((byte[]) (message.getPayload()));
```

Comment:
See comment here: https://github.com/spring-cloud/spring-cloud-stream/pull/598#discussion_r71621110

We can do an assert, but a message bearing this specific content type header which has not a byte[] payload is an anomaly.